App crashes due to permission issue.

App crashes after clicking on Expenses tab.This is after the new update.
Tried a fresh install too.

Error Log from logcat:
```
 FATAL EXCEPTION: main
                                                   Process: com.ae.apps.tripmeter, PID: 11111
                                                   java.lang.RuntimeException: Unable to start activity ComponentInfo{com.ae.apps.tripmeter/com.ae.apps.tripmeter.activities.MainActivity}: java.lang.SecurityException: Permission Denial: opening provider com.android.providers.contacts.ContactsProvider2 from ProcessRecord{8d29387 11111:com.ae.apps.tripmeter/u0a182} (pid=11111, uid=10182) requires android.permission.READ_CONTACTS or android.permission.WRITE_CONTACTS
                                                       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2423)
                                                       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2483)
                                                       at android.app.ActivityThread.access$900(ActivityThread.java:153)
                                                       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1349)
                                                       at android.os.Handler.dispatchMessage(Handler.java:102)
                                                       at android.os.Looper.loop(Looper.java:148)
                                                       at android.app.ActivityThread.main(ActivityThread.java:5441)
                                                       at java.lang.reflect.Method.invoke(Native Method)
                                                       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:738)
                                                       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:628)
                                                    Caused by: java.lang.SecurityException: Permission Denial: opening provider com.android.providers.contacts.ContactsProvider2 from ProcessRecord{8d29387 11111:com.ae.apps.tripmeter/u0a182} (pid=11111, uid=10182) requires android.permission.READ_CONTACTS or android.permission.WRITE_CONTACTS
                                                       at android.os.Parcel.readException(Parcel.java:1620)
                                                       at android.os.Parcel.readException(Parcel.java:1573)
                                                       at android.app.ActivityManagerProxy.getContentProvider(ActivityManagerNative.java:3553)
                                                       at android.app.ActivityThread.acquireProvider(ActivityThread.java:4797)
                                                       at android.app.ContextImpl$ApplicationContentResolver.acquireUnstableProvider(ContextImpl.java:2031)
                                                       at android.content.ContentResolver.acquireUnstableProvider(ContentResolver.java:1484)
                                                       at android.content.ContentResolver.query(ContentResolver.java:477)
                                                       at android.content.ContentResolver.query(ContentResolver.java:435)
                                                       at com.ae.apps.common.c.a.a(Unknown Source)
                                                       at com.ae.apps.common.c.a.<init>(Unknown Source)
                                                       at com.ae.apps.tripmeter.d.a.<init>(Unknown Source)
                                                       at com.ae.apps.tripmeter.d.b.<init>(Unknown Source)
                                                       at com.ae.apps.tripmeter.d.b.a(Unknown Source)
                                                       at com.ae.apps.tripmeter.b.h.a(Unknown Source)
                                                       at android.support.v4.b.l.j(Unknown Source)
                                                       at android.support.v4.b.r.a(Unknown Source)
                                                       at android.support.v4.b.r.z(Unknown Source)
                                                       at android.support.v4.b.r.a(Unknown Source)
                                                       at android.support.v4.b.r.b(Unknown Source)
                                                       at android.support.v4.b.r.f(Unknown Source)
                                                       at android.support.v4.b.o.n(Unknown Source)
                                                       at android.support.v4.b.m.onStart(Unknown Source)
                                                       at android.support.v7.app.c.onStart(Unknown Source)
                                                       at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1238)
                                                       at android.app.Activity.performStart(Activity.java:6320)
                                                       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2386)
                                                       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2483) 
                                                       at android.app.ActivityThread.access$900(ActivityThread.java:153) 
                                                       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1349) 
                                                       at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                       at android.os.Looper.loop(Looper.java:148) 
                                                       at android.app.ActivityThread.main(ActivityThread.java:5441) 
                                                       at java.lang.reflect.Method.invoke(Native Method) 
                                                       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:738) 
                                                       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:628)
```

--- a/app/src/main/java/com/ae/apps/tripmeter/fragments/TripsListFragment.java
+++ b/app/src/main/java/com/ae/apps/tripmeter/fragments/TripsListFragment.java
@@ -29,7 +29,6 @@ import android.support.design.widget.FloatingActionButton;
 import android.support.v4.app.ActivityCompat;
 import android.support.v4.app.Fragment;
 import android.support.v4.app.FragmentManager;
-import android.support.v4.app.FragmentTransaction;
 import android.support.v4.content.ContextCompat;
 import android.support.v7.app.AlertDialog;
 import android.support.v7.widget.LinearLayoutManager;
@@ -38,6 +37,7 @@ import android.util.Log;
 import android.view.LayoutInflater;
 import android.view.View;
 import android.view.ViewGroup;
+import android.widget.TextView;
 import android.widget.Toast;
 
 import com.ae.apps.common.vo.ContactVo;
@@ -72,6 +72,10 @@ public class TripsListFragment extends Fragment
 
     private List<Trip> mTrips;
 
+    private LayoutInflater mInflater;
+
+    private ViewGroup mContainer;
+
     public TripsListFragment() {
     }
 
@@ -87,54 +91,29 @@ public class TripsListFragment extends Fragment
     @Override
     public View onCreateView(LayoutInflater inflater, ViewGroup container,
                              Bundle savedInstanceState) {
-        View view = inflater.inflate(R.layout.fragment_trips_list, container, false);
-        View list = view.findViewById(R.id.list);
+        mInflater = inflater;
+        mContainer = container;
 
+        View placeholder =inflater.inflate(R.layout.fragment_trip_empty,container,false);
+        checkPermissions();
+        return placeholder;
         // Expenses needs default profile to be set inorder to function
         // Check if one has been selected or ask for 1 to be selected
         // Removing default profile feature
         // checkForDefaultProfile();
+    }
 
-        view=createViewHandlingPermissions(view,list);
-
+    private View createNoAccessView(){
+        View view =mInflater.inflate(R.layout.fragment_trip_empty,mContainer,false);
+        TextView noContactsAccess = (TextView) view.findViewById(R.id.contacts_access);
+        noContactsAccess.setVisibility(View.VISIBLE);
         return view;
     }
 
-    private View createViewHandlingPermissions(View view,View list){
-        if (ContextCompat.checkSelfPermission(getActivity(),
-                Manifest.permission.READ_CONTACTS)
-                != PackageManager.PERMISSION_GRANTED) {
-            if (ActivityCompat.shouldShowRequestPermissionRationale(getActivity(),
-                    Manifest.permission.READ_CONTACTS)) {
-                AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
-                builder.setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
-                    @Override
-                    public void onClick(DialogInterface dialog, int which) {
-                        requestPermissions(new String[]{Manifest.permission.READ_CONTACTS},
-                                PERMISSIONS_REQUEST_READ_CONTACTS);
-                        dialog.dismiss();
-                    }
-                })
-                        .setNegativeButton(android.R.string.no, new DialogInterface.OnClickListener() {
-                            @Override
-                            public void onClick(DialogInterface dialog, int which) {
-                                dialog.dismiss();
-                            }
-                        })
-                        .setMessage(R.string.str_perm_reason)
-                        .setTitle(R.string.str_perm_title);
-                builder.create().show();
-            } else {
-                requestPermissions(new String[]{Manifest.permission.READ_CONTACTS},
-                        PERMISSIONS_REQUEST_READ_CONTACTS);
-            }
-        }else{
-            view=creatingView(view,list);
-        }
-        return view;
-    }
+    private View createExpenseView(){
+        View view = mInflater.inflate(R.layout.fragment_trips_list, mContainer, false);
+        View list = view.findViewById(R.id.list);
 
-    private View creatingView(View view,View list){
         mExpenseManager = ExpenseManager.newInstance(getActivity());
         mTrips = mExpenseManager.getAllTrips();
         mViewAdapter = new TripRecyclerViewAdapter(mTrips, mListener);
@@ -157,6 +136,39 @@ public class TripsListFragment extends Fragment
         return view;
     }
 
+    private void checkPermissions(){
+        if (ContextCompat.checkSelfPermission(getActivity(),
+                Manifest.permission.READ_CONTACTS)
+                != PackageManager.PERMISSION_GRANTED) {
+            if (ActivityCompat.shouldShowRequestPermissionRationale(getActivity(),
+                    Manifest.permission.READ_CONTACTS)) {
+                AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
+                builder.setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
+                    @Override
+                    public void onClick(DialogInterface dialog, int which) {
+                        requestPermissions(new String[]{Manifest.permission.READ_CONTACTS},
+                                PERMISSIONS_REQUEST_READ_CONTACTS);
+                        dialog.dismiss();
+                    }
+                })
+                        .setNegativeButton(android.R.string.no, new DialogInterface.OnClickListener() {
+                            @Override
+                            public void onClick(DialogInterface dialog, int which) {
+                                dialog.dismiss();
+                            }
+                        })
+                        .setMessage(R.string.str_permission_contacts_reason)
+                        .setTitle(R.string.str_permission_title);
+                builder.create().show();
+            } else {
+                requestPermissions(new String[]{Manifest.permission.READ_CONTACTS},
+                        PERMISSIONS_REQUEST_READ_CONTACTS);
+            }
+        }else{
+            View view=createExpenseView();
+            inflateView(view);
+        }
+    }
 
     @Override
     public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
@@ -164,10 +176,12 @@ public class TripsListFragment extends Fragment
             case PERMISSIONS_REQUEST_READ_CONTACTS: {
                 if (grantResults.length > 0
                         && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
-                    //TODO: Reload this Fragment.Otherwise the FAB won't work
+                    View view=createExpenseView();
+                    inflateView(view);
                 } else {
-                    Toast.makeText(getActivity(), "This feature needs access to contacts.", Toast.LENGTH_LONG).show();
-                    //TODO: Direct user to price fragment as this fragment cannot be used
+                    Toast.makeText(getActivity(), R.string.str_permission_contacts_reason, Toast.LENGTH_LONG).show();
+                    View view=createNoAccessView();
+                    inflateView(view);
                 }
                 return;
             }
@@ -176,6 +190,11 @@ public class TripsListFragment extends Fragment
                 return;
             }
         }
+    }
+
+    private void inflateView(View view){
+        mContainer.removeAllViews();
+        mContainer.addView(view);
     }
 
     @SuppressWarnings("unused")

--- a/app/src/main/java/com/ae/apps/tripmeter/fragments/TripsListFragment.java
+++ b/app/src/main/java/com/ae/apps/tripmeter/fragments/TripsListFragment.java
@@ -19,17 +19,26 @@
  */
 package com.ae.apps.tripmeter.fragments;
 
+import android.Manifest;
 import android.content.Context;
+import android.content.DialogInterface;
+import android.content.pm.PackageManager;
 import android.os.Bundle;
+import android.support.annotation.NonNull;
 import android.support.design.widget.FloatingActionButton;
+import android.support.v4.app.ActivityCompat;
 import android.support.v4.app.Fragment;
 import android.support.v4.app.FragmentManager;
+import android.support.v4.app.FragmentTransaction;
+import android.support.v4.content.ContextCompat;
+import android.support.v7.app.AlertDialog;
 import android.support.v7.widget.LinearLayoutManager;
 import android.support.v7.widget.RecyclerView;
 import android.util.Log;
 import android.view.LayoutInflater;
 import android.view.View;
 import android.view.ViewGroup;
+import android.widget.Toast;
 
 import com.ae.apps.common.vo.ContactVo;
 import com.ae.apps.tripmeter.R;
@@ -53,6 +62,8 @@ public class TripsListFragment extends Fragment
 
     private static final String TAG = "TripsListFragment";
 
+    private static final int PERMISSIONS_REQUEST_READ_CONTACTS=1;
+
     private ExpensesInteractionListener mListener;
 
     private ExpenseManager mExpenseManager;
@@ -71,7 +82,6 @@ public class TripsListFragment extends Fragment
     @Override
     public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
-        mExpenseManager = ExpenseManager.newInstance(getActivity());
     }
 
     @Override
@@ -85,6 +95,47 @@ public class TripsListFragment extends Fragment
         // Removing default profile feature
         // checkForDefaultProfile();
 
+        view=createViewHandlingPermissions(view,list);
+
+        return view;
+    }
+
+    private View createViewHandlingPermissions(View view,View list){
+        if (ContextCompat.checkSelfPermission(getActivity(),
+                Manifest.permission.READ_CONTACTS)
+                != PackageManager.PERMISSION_GRANTED) {
+            if (ActivityCompat.shouldShowRequestPermissionRationale(getActivity(),
+                    Manifest.permission.READ_CONTACTS)) {
+                AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
+                builder.setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
+                    @Override
+                    public void onClick(DialogInterface dialog, int which) {
+                        requestPermissions(new String[]{Manifest.permission.READ_CONTACTS},
+                                PERMISSIONS_REQUEST_READ_CONTACTS);
+                        dialog.dismiss();
+                    }
+                })
+                        .setNegativeButton(android.R.string.no, new DialogInterface.OnClickListener() {
+                            @Override
+                            public void onClick(DialogInterface dialog, int which) {
+                                dialog.dismiss();
+                            }
+                        })
+                        .setMessage(R.string.str_perm_reason)
+                        .setTitle(R.string.str_perm_title);
+                builder.create().show();
+            } else {
+                requestPermissions(new String[]{Manifest.permission.READ_CONTACTS},
+                        PERMISSIONS_REQUEST_READ_CONTACTS);
+            }
+        }else{
+            view=creatingView(view,list);
+        }
+        return view;
+    }
+
+    private View creatingView(View view,View list){
+        mExpenseManager = ExpenseManager.newInstance(getActivity());
         mTrips = mExpenseManager.getAllTrips();
         mViewAdapter = new TripRecyclerViewAdapter(mTrips, mListener);
 
@@ -103,8 +154,28 @@ public class TripsListFragment extends Fragment
                 showAddTripDialog();
             }
         });
-
         return view;
+    }
+
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+        switch (requestCode) {
+            case PERMISSIONS_REQUEST_READ_CONTACTS: {
+                if (grantResults.length > 0
+                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                    //TODO: Reload this Fragment.Otherwise the FAB won't work
+                } else {
+                    Toast.makeText(getActivity(), "This feature needs access to contacts.", Toast.LENGTH_LONG).show();
+                    //TODO: Direct user to price fragment as this fragment cannot be used
+                }
+                return;
+            }
+            default:{
+                super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+                return;
+            }
+        }
     }
 
     @SuppressWarnings("unused")

--- a/app/src/main/res/layout/fragment_trip_empty.xml
+++ b/app/src/main/res/layout/fragment_trip_empty.xml
@@ -1,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
+<android.support.design.widget.CoordinatorLayout
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:layout_gravity="center"
+    tools:context="com.ae.apps.tripmeter.fragments.TripsListFragment">
+
+    <TextView
+        android:id="@+id/contacts_access"
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:layout_gravity="center_vertical"
+        android:drawablePadding="@dimen/fab_margin"
+        android:gravity="center"
+        android:text="@string/str_permission_contacts_reason"
+        android:visibility="gone" />
+
+</android.support.design.widget.CoordinatorLayout>

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -134,6 +134,9 @@
 
     <string name="pref_title_vibrate">Vibrate</string>
 
+    <string name="str_perm_title">Permission Needed</string>
+    <string name="str_perm_reason">Contacts access is needed for this feature</string>
+
     <!-- TODO: Remove or change this placeholder text -->
     <string name="hello_blank_fragment">Hello blank fragment</string>
 

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -134,8 +134,8 @@
 
     <string name="pref_title_vibrate">Vibrate</string>
 
-    <string name="str_perm_title">Permission Needed</string>
-    <string name="str_perm_reason">Contacts access is needed for this feature</string>
+    <string name="str_permission_title">Permission Needed</string>
+    <string name="str_permission_contacts_reason">Contacts access is needed for this feature</string>
 
     <!-- TODO: Remove or change this placeholder text -->
     <string name="hello_blank_fragment">Hello blank fragment</string>

